src/js/trove/matrix.js

```diff
@@ -0,0 +1,1310 @@
+        var fillMatrix = function (h, w, n) {
+            arity(3, arguments, "fill-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "fill-mat", h, runtime.Number, w, runtime.Number, n, runtime.Number);
+            return createMatrixFromArray(h, w, new Array(h * w).fill(n));
+        }
+
+        /**
+         * 
+         * @param {Number} n 
+         * @returns Identity Matrix of dimensions nxn
+         */
+        var IdentityMatrix = function (n) {
+            arity(1, arguments, "identity-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "identity-mat", n, runtime.Number);
+            if (n < 0) {
+                runtime.ffi.throwMessageException("Negative Dimension");
+            } else {
+                var req_mat = fillMatrix(n, n, 0);
+                for (var i = 0; i < n; i++) {
+                    req_mat.$underlyingMat[get1dpos(i, i, n)] = 1;
+                }
+                return req_mat;
+            }
+        }
+        // Is the matrix a row matrix i.e has dimensions (n,1)
+        var isRowMatrix = function (self) {
+            arity(1, arguments, "is-row-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "is-row-mat", self, annMatrix);
+            if (self.$h == 1) {
+                return runtime.makeBoolean(true);
+            } else {
+                return runtime.makeBoolean(false);
+            }
+        }
+        // Is the matrix a col matrix i.e has dimensions (1,n)
+        var isColMatrix = function (self) {
+            arity(1, arguments, "is-col-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "is-col-mat", self, annMatrix);
+            if (self.$w == 1) {
+                return runtime.makeBoolean(true);
+            } else {
+                return runtime.makeBoolean(false);
+            }
+        }
+        //Is the matrix square
+        var isSquareMatrix = function (self) {
+            arity(1, arguments, "is-square-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "is-square-mat", self, annMatrix);
+            if (self.$w == self.$h) {
+                return runtime.makeBoolean(true);
+            } else {
+                return runtime.makeBoolean(false);
+            }
+        }
+        //Calculates trace of matrix
+        var traceMatrix = function (self) {
+            arity(1, arguments, "trace-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "trace-mat", self, annMatrix);
+            var res = runtime.makeNumber(0);
+            if (!isSquareMatrix(self)) {
+                return runtime.ffi.throwMessageException("Matrix must be square");
+            }
+            for (var i = 0; i < self.$h; i++) {
+                res  = runtime.plus(res,self.$underlyingMat[get1dpos(i, i, self.$w)]);
+            }
+            return res;
+        }
+
+        //Returns the Reduced Row Echelon Form of the Matrix
+        var MatRref = function (self) {
```

Comment:
Why is this `MatRref` and not `matrixReducedRowEchelon` or some other more standard-looking name?  (Capitalized names, certainly, indicate object constructors, not functions...)